lib/paper_trail/model_config.rb

```diff
@@ -147,20 +139,31 @@ def cannot_record_after_destroy?
     # is "Dog". If `attrs["species"]` is blank, `item_type` is "Animal". See
     # `spec/models/animal_spec.rb`.
     def setup_versions_association(klass)
-      klass.has_many(
+      has_manys = klass.has_many(
         klass.versions_association_name,
         lambda do
-          relation = order(model.timestamp_sort_order)
-          item_type = klass.paper_trail.send(:versions_association_item_type)
-          relation = relation.unscope(where: :item_type).where(item_type: item_type)
-          relation
+          order!(model.timestamp_sort_order)
+          unscope(where: :item_type).where(item_type: klass.name) unless klass.descendants.any?
         end,
         class_name: klass.version_class_name,
         as: :item
       )
 
-      # We override the assocation when STI models are created from a base class
-      # in order to use the right `item_type`.
+      # Only for this .versions association override HasManyAssociation#collection?
+      # (that normally always returns true) so it returns false when referring to
+      # a subclassed model that uses STI.  Allows .create() events not to revert
+      # back to base_class at the final stages when Association#creation_attributes
+      # gets called.
+      has_manys[klass.versions_association_name.to_s].define_singleton_method(:collection?) do
+        active_record.descends_from_active_record?
```

Comment:
How confident can we be that `collection?` will continue to be used by Rails in this way? Why does Rails have this method in the first place, just to return true?

Reply:
How often does `collection?` end up getting called? I wonder if it's worth the effort to memoize the result of `descends_from_active_record?` so it's not calculated all the time.

BTW, when monkeypatching code I like to add a link to the version-specific lines of code like this: https://github.com/rails/rails/blob/v5.2.1/activerecord/lib/active_record/associations/association.rb#L198-L210

Since `creation_attributes` only sets the foreign key & type attributes, wouldn't it be more expressive to override it directly?
```ruby
has_manys[klass.versions_association_name.to_s].singleton_class.prepend(Module.new {
  def creation_attributes
    if active_record.descends_from_active_record?
      super # sets `item_type` and `item_id`
    else
      {} # we set these elsewhere. Rails would incorrectly set them to the STI base class.
    end
  end
})
```

That raises the question: why don't we always pass in the correct `item_type` via this monkeypatch? If we're going to patch Rails, it might as well be consistent.

Reply:
CC @jaredbeck

Reply:
> How confident can we be that `collection?` will continue to be used by Rails in this way?

It's been used in this way since Rails v4.0, and before that a very similar test was used, albeit one that wouldn't work out with our override:

```
if reflection.macro.in?([:has_one, :has_many]) && !options[:through]
```

> Since `creation_attributes` only sets the foreign key & type attributes, wouldn't it be more expressive to override it directly?

At first I had hoped to have a solution that utilised this approach ... but it requires a little more code.  The only way I see it working is as a class method added to the HasManyAssociation class, and inside a test to see if we're dealing with a PaperTrail::Version.  For fun I've put that together in the latest commit.  10 lines of stuff as opposed to this little 4-liner.

I'm certainly OK with having this either way -- both approaches establish the same net result.

> why don't we always pass in the correct item_type via this monkeypatch?

It's very specific just to PaperTrail::Version objects, and only has effect for anything subclassed.

Reply:
Re: https://github.com/paper-trail-gem/paper_trail/pull/1137#discussion_r211112725

Are we okay with globally patching `creation_attributes`? Additionally, if we are okay with it, should we perhaps only patch it once instead of patching it every time `has_paper_trail` is called?

Reply:
Good points on `creation_attributes` -- have gone back to directly patching `collection?` on the specific :verisons `HasManyReflection` objects that are actually subclassed.  Much less intrusive.